internal/resources/providers/gcplib/inventory/provider.go

```diff
@@ -297,41 +298,101 @@ func (p *Provider) getAssetAncestorsPolicies(ctx context.Context, ancestors []st
 }
 
 func (p *Provider) getParentResources(ctx context.Context, parent string, assetTypes []string) <-chan *ExtendedGcpAsset {
+	return p.getAssets(ctx, parent, assetTypes, assetpb.ContentType_RESOURCE)
+}
+
+func (p *Provider) getAssets(ctx context.Context, parent string, assetTypes []string, contentType assetpb.ContentType) <-chan *ExtendedGcpAsset {
 	ch := make(chan *ExtendedGcpAsset)
 	go p.getAllAssets(ctx, ch, &assetpb.ListAssetsRequest{
 		Parent:      parent,
 		AssetTypes:  assetTypes,
-		ContentType: assetpb.ContentType_RESOURCE,
+		ContentType: contentType,
 	})
 	return ch
 }
 
-func (p *Provider) fetchAssets(ctx context.Context, contentType assetpb.ContentType, assetTypes []string) <-chan *ExtendedGcpAsset {
+func (p *Provider) mergeAssets(ctx context.Context, resCh, polCh <-chan *ExtendedGcpAsset) <-chan *ExtendedGcpAsset {
 	out := make(chan *ExtendedGcpAsset)
-	wg := sync.WaitGroup{}
-	// Fetch each asset type separately to limit failures to a single type
-	for _, assetType := range assetTypes {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			ch := make(chan *ExtendedGcpAsset)
-			go p.getAllAssets(ctx, ch, &assetpb.ListAssetsRequest{
-				Parent:      p.config.Parent,
-				AssetTypes:  []string{assetType},
-				ContentType: contentType,
-			})
-			for asset := range ch {
-				out <- asset
-			}
-		}()
-	}
+	store := make(map[string]*ExtendedGcpAsset)
+
 	go func() {
-		wg.Wait()
-		close(out)
+		defer close(out)
+
+		for polCh != nil || resCh != nil {
+			select {
+			case <-ctx.Done():
+				return
+
+			case a, ok := <-polCh:
+				if !ok {
+					polCh = nil
+					flushAssets(out, store, resCh, polCh)
```

Comment:
Isn't a single `flushAssets` call when both channels are closed is sufficient? we already check in `mergeAsset` if one of the channels is already closed in order to expedite processing.

Reply:
for correctness - yeah, we could do with a single `flushAssets`. but for performance - we shouldn't, as these 2 additional calls mean we drain the store faster which directly impacts memory usage.

without it, after one channel closes (let's assume policies channel closed), `mergeAsset` will only send **the next** asset that comes because 1) it found its policy pair, or 2) it didn't but the policies channel is closed so we don't need to wait. but what about all the resources already in the map? by now we know they don't have a policy, and if we don't flush, we'll have to wait until the resource channel is closed too.